src/common/decorators/one-at-time.ts

```diff
@@ -1,22 +1,71 @@
+const stakingModuleId = Symbol('StakingModuleId');
+
+export function StakingModuleId(
+  target: any,
+  propertyKey: string | symbol,
+  parameterIndex: number,
+) {
+  const existingMetadata: number[] =
+    Reflect.getOwnMetadata(stakingModuleId, target, propertyKey) || [];
+
+  if (existingMetadata.length === 0) {
+    Reflect.defineMetadata(
+      stakingModuleId,
+      [parameterIndex],
+      target,
+      propertyKey,
+    );
+  } else {
+    throw new Error(
+      `StakingModuleId decorator can only be applied to one parameter in method ${String(
+        propertyKey,
+      )}. It is already applied to parameter index ${existingMetadata[0]}`,
+    );
+  }
+}
+
 export function OneAtTime<T extends (...args: any[]) => Promise<any>>() {
   return function (
-    target: unknown,
+    target: any,
     propertyName: string,
     descriptor: TypedPropertyDescriptor<T>,
   ) {
     const method = descriptor.value;
     let isExecuting = false;
+    const isExecutingMap = new Map<number, boolean>();
 
     descriptor.value = async function (this: any, ...args) {
-      if (isExecuting) return;
+      const stakingModuleIdArgs =
+        Reflect.getMetadata(stakingModuleId, target, propertyName) || [];
+
+      const moduleId =
+        stakingModuleIdArgs.length > 0 ? args[stakingModuleIdArgs[0]] : null;
```

Comment:
So if function has any input params, module id must be in first place?

Reply:
not exactly
   const stakingModuleIdArgs =
        Reflect.getMetadata(stakingModuleId, target, propertyName) || [];

We fetch here only arguments with stakingModuleId decorator. But in StakingModuleId we check that in function can be only one parameter.